File: go-backend-interview-prep/10_design_patterns_in_go/07_observer_events/main.go
package main

import "fmt"

// --- Event types ---

type Event struct {
	Name string
	Data map[string]string
}

// --- Event bus ---

type EventBus struct {
	handlers map[string][]func(Event)
}

func NewEventBus() *EventBus {
	return &EventBus{handlers: make(map[string][]func(Event))}
}

func (b *EventBus) Subscribe(eventName string, handler func(Event)) {
	b.handlers[eventName] = append(b.handlers[eventName], handler)
}

func (b *EventBus) Publish(event Event) {
	handlers, ok := b.handlers[event.Name]
	if !ok {
		return
	}
	for _, h := range handlers {
		h(event) // in production: run in goroutine with recover
	}
}

// --- Notification service publishes events ---

type NotificationService struct {
	bus *EventBus
}

func NewNotificationService(bus *EventBus) *NotificationService {
	return &NotificationService{bus: bus}
}

func (s *NotificationService) Send(to, message string) {
	fmt.Printf("  [SEND] To: %s | %s\n", to, message)
	s.bus.Publish(Event{
		Name: "notification.sent",
		Data: map[string]string{"to": to, "message": message},
	})
}

func (s *NotificationService) Fail(to, reason string) {
	fmt.Printf("  [FAIL] To: %s | reason: %s\n", to, reason)
	s.bus.Publish(Event{
		Name: "notification.failed",
		Data: map[string]string{"to": to, "reason": reason},
	})
}

func main() {
	fmt.Println("=== Observer / Events Pattern ===")
	fmt.Println()

	bus := NewEventBus()

	// Subscribe: logger
	bus.Subscribe("notification.sent", func(e Event) {
		fmt.Printf("  [LOG]       Notification sent to %s\n", e.Data["to"])
	})

	// Subscribe: analytics
	bus.Subscribe("notification.sent", func(e Event) {
		fmt.Printf("  [ANALYTICS] Recorded delivery to %s\n", e.Data["to"])
	})

	// Subscribe: alert on failure
	bus.Subscribe("notification.failed", func(e Event) {
		fmt.Printf("  [ALERT]     FAILURE for %s: %s\n", e.Data["to"], e.Data["reason"])
	})

	// --- Use the service ---
	svc := NewNotificationService(bus)

	fmt.Println("--- Sending notifications ---")
	svc.Send("alice@example.com", "Order confirmed")
	fmt.Println()
	svc.Send("bob@example.com", "Payment received")
	fmt.Println()

	fmt.Println("--- Simulating failure ---")
	svc.Fail("charlie@example.com", "invalid email")
	fmt.Println()

	fmt.Println("Key: service publishes events. Subscribers react independently.")
	fmt.Println("     Adding analytics = one Subscribe() call. Zero service changes.")
}

File: go-backend-interview-prep/10_design_patterns_in_go/07_observer_events/main.py
"""Observer / events pattern -- Python equivalent of the Go example.

Simple event bus with subscribe/publish.
"""

from dataclasses import dataclass, field


@dataclass
class Event:
    name: str
    data: dict = field(default_factory=dict)


class EventBus:
    def __init__(self):
        self.handlers: dict[str, list] = {}

    def subscribe(self, event_name: str, handler) -> None:
        self.handlers.setdefault(event_name, []).append(handler)

    def publish(self, event: Event) -> None:
        for handler in self.handlers.get(event.name, []):
            handler(event)


class NotificationService:
    def __init__(self, bus: EventBus):
        self.bus = bus

    def send(self, to: str, message: str) -> None:
        print(f"  [SEND] To: {to} | {message}")
        self.bus.publish(Event("notification.sent", {"to": to, "message": message}))

    def fail(self, to: str, reason: str) -> None:
        print(f"  [FAIL] To: {to} | reason: {reason}")
        self.bus.publish(Event("notification.failed", {"to": to, "reason": reason}))


def main() -> None:
    print("=== Observer / Events Pattern ===")
    print()

    bus = EventBus()

    # Subscribe: logger
    bus.subscribe("notification.sent",
                  lambda e: print(f"  [LOG]       Notification sent to {e.data['to']}"))

    # Subscribe: analytics
    bus.subscribe("notification.sent",
                  lambda e: print(f"  [ANALYTICS] Recorded delivery to {e.data['to']}"))

    # Subscribe: alert on failure
    bus.subscribe("notification.failed",
                  lambda e: print(f"  [ALERT]     FAILURE for {e.data['to']}: {e.data['reason']}"))

    svc = NotificationService(bus)

    print("--- Sending notifications ---")
    svc.send("alice@example.com", "Order confirmed")
    print()
    svc.send("bob@example.com", "Payment received")
    print()

    print("--- Simulating failure ---")
    svc.fail("charlie@example.com", "invalid email")
    print()

    print("Key: service publishes events. Subscribers react independently.")
    print("     Adding analytics = one subscribe() call. Zero service changes.")


if __name__ == "__main__":
    main()
